File: v3analyzer/parser.py
"""
Parser for Paradox PDXScript text format.

Uses regex-based tokenization for performance (C-speed), then builds
the tree in a single pass.

Handles:
- key = value pairs
- key = { ... } nested blocks
- Arrays (lists of values inside { })
- Quoted and unquoted strings
- Numbers (int and float)
- Dates like 1836.1.1
- yes/no booleans
- Duplicate keys (stored as lists)
- Operators: = > < >= <=
"""
import re
import sys
from typing import Any

# Increase recursion limit for deeply nested save files
sys.setrecursionlimit(15000)

# Single regex that tokenizes all PDX elements.
# Ordering matters: comments and quoted strings must match before operators/tokens.
_TOKEN_RE = re.compile(r"""
    \#[^\n]*             |  # comment — skip entirely
    ("(?:[^"\\]|\\.)*")  |  # quoted string (group 1)
    ([{}])               |  # braces (group 2)
    ([=><]=?)            |  # operators (group 3)
    ([^\s={}><\#"]+)        # unquoted token (group 4)
""", re.VERBOSE)


def _convert(token: str) -> Any:
    """Convert an unquoted string token to the appropriate Python type."""
    if token == "yes":
        return True
    if token == "no":
        return False
    try:
        return int(token)
    except ValueError:
        pass
    try:
        return float(token)
    except ValueError:
        pass
    return token


def _tokenize(text: str) -> list:
    """Tokenize PDXScript text into a flat list of (type, value) tuples.

    Types: 's' = string, 'b' = brace, 'o' = operator, 't' = token
    Comments are discarded. This runs at C speed via re.finditer.
    """
    tokens = []
    append = tokens.append  # local ref for speed
    for m in _TOKEN_RE.finditer(text):
        g1 = m.group(1)
        if g1 is not None:
            append(('s', g1[1:-1]))  # strip quotes
            continue
        g2 = m.group(2)
        if g2 is not None:
            append(('b', g2))
            continue
        g3 = m.group(3)
        if g3 is not None:
            append(('o', g3))
            continue
        g4 = m.group(4)
        if g4 is not None:
            append(('t', g4))
    return tokens


class PDXParser:
    """Parse PDXScript text into Python data structures.

    Uses fast regex tokenization followed by a single-pass recursive
    descent tree builder.
    """

    def __init__(self, text: str):
        # Strip SAV header line if present (Rakaly melted saves)
        if text.startswith('SAV'):
            nl = text.index('\n')
            text = text[nl + 1:]
        self._tokens = _tokenize(text)
        self._n = len(self._tokens)
        self._pos = 0

    def parse(self) -> dict:
        """Parse the entire text and return a dict."""
        return self._parse_pairs()

    # ---- internal recursive descent on token list ----

    def _parse_pairs(self) -> dict:
        """Parse key=value pairs until } or end of tokens."""
        result = {}
        tokens = self._tokens
        n = self._n
        while self._pos < n:
            typ, val = tokens[self._pos]
            if typ == 'b' and val == '}':
                return result

            self._pos += 1

            # Expect operator next for a key=value pair
            if self._pos < n and tokens[self._pos][0] == 'o':
                self._pos += 1  # consume operator
                pval = self._parse_value()
                key = val
                if key in result:
                    existing = result[key]
                    if isinstance(existing, list):
                        existing.append(pval)
                    else:
                        result[key] = [existing, pval]
                else:
                    result[key] = pval
            else:
                # No operator — bare value; backtrack so caller can handle
                self._pos -= 1
                return result

        return result

    def _parse_value(self) -> Any:
        """Parse a single value: block, string, or primitive token."""
        if self._pos >= self._n:
            return ""
        typ, val = self._tokens[self._pos]
        if typ == 'b' and val == '{':
            return self._parse_block()
        self._pos += 1
        if typ == 's':
            return val
        # Handle color literals: rgb { R G B } or hsv { H S V }
        converted = _convert(val)
        if val in ('rgb', 'hsv') and self._pos < self._n:
            ntyp, nval = self._tokens[self._pos]
            if ntyp == 'b' and nval == '{':
                block = self._parse_block()
                return val  # discard color values, keep as string label
        return converted

    def _parse_block(self) -> Any:
        """Parse a { ... } block as either a dict or array."""
        self._pos += 1  # skip '{'
        if self._pos >= self._n:
            return {}
        typ, val = self._tokens[self._pos]
        if typ == 'b' and val == '}':
            self._pos += 1
            return {}

        # Peek: if second token is an operator, parse as dict
        if self._pos + 1 < self._n and self._tokens[self._pos + 1][0] == 'o':
            result = self._parse_pairs()
        else:
            result = self._parse_array()

        # Consume closing brace
        if (self._pos < self._n
                and self._tokens[self._pos][0] == 'b'
                and self._tokens[self._pos][1] == '}'):
            self._pos += 1
        return result

    def _parse_array(self) -> Any:
        """Parse an array of values inside { }.

        If a key=value pair is encountered mid-array, switches to dict mode.
        """
        result = []
        tokens = self._tokens
        n = self._n
        while self._pos < n:
            typ, val = tokens[self._pos]
            if typ == 'b' and val == '}':
                return result
            # Detect key=value → switch to dict mode
            if self._pos + 1 < n and tokens[self._pos + 1][0] == 'o':
                d = self._parse_pairs()
                if not result:
                    return d
                # Rare: array followed by key=value pairs; return array
                return result
            result.append(self._parse_value())
        return result


def parse_pdx(text: str) -> dict:
    """Convenience function to parse PDXScript text."""
    parser = PDXParser(text)
    return parser.parse()

File: web/src/parser.ts
/**
 * Parser for Paradox PDXScript text format.
 *
 * Uses regex-based tokenization then builds the tree in a single pass.
 * Handles: key=value pairs, nested blocks, arrays, quoted/unquoted strings,
 * numbers, dates (as strings), yes/no booleans, duplicate keys (→ arrays),
 * operators: = > < >= <=
 */

import type { ParsedData, ParsedValue } from './types';

const enum TokenKind {
  String,
  Brace,
  Operator,
  Text,
}

interface Token {
  kind: TokenKind;
  value: string;
}

const TOKEN_RE =
  /\#[^\n]*|("(?:[^"\\]|\\.)*")|([{}])|([=><]=?)|([^\s={}><\#"]+)/g;

function coerce(raw: string): string | number | boolean {
  if (raw === 'yes') return true;
  if (raw === 'no') return false;
  const asInt = parseInt(raw, 10);
  if (String(asInt) === raw) return asInt;
  const asFloat = parseFloat(raw);
  if (!isNaN(asFloat) && /^-?\d+\.\d+$/.test(raw)) return asFloat;
  return raw;
}

function tokenize(text: string): Token[] {
  const tokens: Token[] = [];
  let m: RegExpExecArray | null;
  TOKEN_RE.lastIndex = 0;
  while ((m = TOKEN_RE.exec(text)) !== null) {
    if (m[1] !== undefined) {
      tokens.push({ kind: TokenKind.String, value: m[1].slice(1, -1) });
    } else if (m[2] !== undefined) {
      tokens.push({ kind: TokenKind.Brace, value: m[2] });
    } else if (m[3] !== undefined) {
      tokens.push({ kind: TokenKind.Operator, value: m[3] });
    } else if (m[4] !== undefined) {
      tokens.push({ kind: TokenKind.Text, value: m[4] });
    }
  }
  return tokens;
}

class PDXParser {
  private readonly tokens: Token[];
  private readonly length: number;
  private pos: number;

  constructor(text: string) {
    // Strip SAV header line if present (Rakaly melted saves)
    if (text.startsWith('SAV')) {
      const newline = text.indexOf('\n');
      text = text.substring(newline + 1);
    }
    this.tokens = tokenize(text);
    this.length = this.tokens.length;
    this.pos = 0;
  }

  parse(): ParsedData {
    return this.parsePairs();
  }

  private parsePairs(): ParsedData {
    const result: ParsedData = {};
    while (this.pos < this.length) {
      const token = this.tokens[this.pos];
      if (token.kind === TokenKind.Brace && token.value === '}') return result;

      this.pos++;

      if (this.pos < this.length && this.tokens[this.pos].kind === TokenKind.Operator) {
        this.pos++; // consume operator
        const value = this.parseValue();
        const key = token.value;
        if (key in result) {
          const existing = result[key];
          if (Array.isArray(existing)) {
            existing.push(value);
          } else {
            result[key] = [existing, value];
          }
        } else {
          result[key] = value;
        }
      } else {
        this.pos--;
        return result;
      }
    }
    return result;
  }

  private parseValue(): ParsedValue {
    if (this.pos >= this.length) return '';
    const token = this.tokens[this.pos];
    if (token.kind === TokenKind.Brace && token.value === '{') return this.parseBlock();
    this.pos++;
    if (token.kind === TokenKind.String) return token.value;
    const converted = coerce(token.value);
    // Handle color literals: rgb { R G B } or hsv { H S V }
    if (
      (token.value === 'rgb' || token.value === 'hsv') &&
      this.pos < this.length
    ) {
      const next = this.tokens[this.pos];
      if (next.kind === TokenKind.Brace && next.value === '{') {
        this.parseBlock(); // discard color values
        return token.value;
      }
    }
    return converted;
  }

  private parseBlock(): ParsedValue {
    this.pos++; // skip '{'
    if (this.pos >= this.length) return {};
    const token = this.tokens[this.pos];
    if (token.kind === TokenKind.Brace && token.value === '}') {
      this.pos++;
      return {};
    }

    let result: ParsedValue;
    if (
      this.pos + 1 < this.length &&
      this.tokens[this.pos + 1].kind === TokenKind.Operator
    ) {
      result = this.parsePairs();
    } else {
      result = this.parseArray();
    }

    // Consume closing brace
    if (
      this.pos < this.length &&
      this.tokens[this.pos].kind === TokenKind.Brace &&
      this.tokens[this.pos].value === '}'
    ) {
      this.pos++;
    }
    return result;
  }

  private parseArray(): ParsedValue {
    const result: ParsedValue[] = [];
    while (this.pos < this.length) {
      const token = this.tokens[this.pos];
      if (token.kind === TokenKind.Brace && token.value === '}') return result;
      // Detect key=value → switch to dict mode
      if (
        this.pos + 1 < this.length &&
        this.tokens[this.pos + 1].kind === TokenKind.Operator
      ) {
        const pairs = this.parsePairs();
        if (result.length === 0) return pairs;
        return result;
      }
      result.push(this.parseValue());
    }
    return result;
  }
}

export function parsePdx(text: string): ParsedData {
  return new PDXParser(text).parse();
}
